common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -460,47 +458,45 @@ public String toString() {
             "io.netty.buffer.AdvancedLeakAwareByteBuf.recordLeakNonRefCountingOperation("
     };
 
-    static String newRecord(Object hint, int recordsToSkip) {
-        StringBuilder buf = new StringBuilder(4096);
+    private static final class Record extends Throwable {
+        private final String hintString;
+        private Record next;
 
-        // Append the hint first if available.
-        if (hint != null) {
-            buf.append("\tHint: ");
-            // Prefer a hint string to a simple string form.
-            if (hint instanceof ResourceLeakHint) {
-                buf.append(((ResourceLeakHint) hint).toHintString());
-            } else {
-                buf.append(hint);
-            }
-            buf.append(NEWLINE);
+        Record(Object hint) {
+            // This needs to be generated even if toString() is never called as it may change later on.
+            hintString = hint instanceof ResourceLeakHint ? ((ResourceLeakHint) hint).toHintString() : hint.toString();
         }
 
-        // Append the stack trace.
-        StackTraceElement[] array = new Throwable().getStackTrace();
-        for (StackTraceElement e: array) {
-            if (recordsToSkip > 0) {
-                recordsToSkip --;
-            } else {
-                String estr = e.toString();
+        Record() {
+           hintString = null;
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder buf = new StringBuilder(4096);
+            if (hintString != null) {
+                buf.append("\tHint: ").append(hintString).append(NEWLINE);
+            }
+
+            // Append the stack trace.
+            StackTraceElement[] array = getStackTrace();
```

Comment:
One other idea for perf:  Rather than construct the string here from the stack trace, why not just pass the stack elements and the hint up?

This would have two benefits:  The critical section in `DefaultResourceLeak` would be much smaller.  Second: it would allocate less garbage.

Reply:
With pass up you mean return it ?

Reply:
So rather than construct and return a String in `Record`, have `DefaultResourceLeak` call `getStackTrace()` and access the hint directly.  i.e. don't use `toString()`.

Reply:
For example ... just construct one big `StringBuilder` and construct it once instead of constructing a string building for each record, and copying the result into another `StringBuilder`.

Reply:
I see that said I not think it really matters as this will only be called if a leak is detected so it should not be called frequently. If its called frequently you have more serious problems to fix imho ;)

Reply:
sound reasonable to me ... low priority IIUC ... are we missing something @carl-mastrangelo?

Reply:
@Scottmitch passing in the string builder sounds best to me, but I suppose this will only get called when there is a problem.

@normanmaurer optional, up to you.

Reply:
I will just ignore this for now and may do another pr if it is a problem later on.